File: src/algorithms/trees/manipulation/flip-equivalent-trees/sources/flip-equivalent-trees_test.go
package main

import "testing"

func makeFETNode(value int, left *BinaryNode, right *BinaryNode) *BinaryNode {
	return &BinaryNode{value: value, left: left, right: right}
}

func fetLeaf(value int) *BinaryNode {
	return &BinaryNode{value: value}
}

func TestFlipEquivalentTreesTwoNulls(t *testing.T) {
	if flipEquivalentTrees(nil, nil) != true {
		t.Error("two nulls should be true")
	}
}

func TestFlipEquivalentTreesOneNull(t *testing.T) {
	if flipEquivalentTrees(fetLeaf(1), nil) != false {
		t.Error("one null should be false")
	}
}

func TestFlipEquivalentTreesIdentical(t *testing.T) {
	treeA := makeFETNode(1, fetLeaf(2), fetLeaf(3))
	treeB := makeFETNode(1, fetLeaf(2), fetLeaf(3))
	if flipEquivalentTrees(treeA, treeB) != true {
		t.Error("identical trees should be true")
	}
}

func TestFlipEquivalentTreesFlippedAtRoot(t *testing.T) {
	treeA := makeFETNode(1, fetLeaf(2), fetLeaf(3))
	treeB := makeFETNode(1, fetLeaf(3), fetLeaf(2))
	if flipEquivalentTrees(treeA, treeB) != true {
		t.Error("flipped at root should be true")
	}
}

func TestFlipEquivalentTreesDifferentRootValues(t *testing.T) {
	if flipEquivalentTrees(fetLeaf(1), fetLeaf(2)) != false {
		t.Error("different root values should be false")
	}
}

func TestFlipEquivalentTreesDifferentLeafValues(t *testing.T) {
	treeA := makeFETNode(1, fetLeaf(2), fetLeaf(3))
	treeB := makeFETNode(1, fetLeaf(9), fetLeaf(3))
	if flipEquivalentTrees(treeA, treeB) != false {
		t.Error("different leaf values should be false")
	}
}

File: src/algorithms/trees/manipulation/flip-equivalent-trees/FlipEquivalentTrees_test.cpp
// g++ -o flip_equiv_test FlipEquivalentTrees_test.cpp && ./flip_equiv_test
#include "sources/FlipEquivalentTrees.cpp"
#include <cassert>
#include <iostream>

BinaryNode* makeFETNode(int value, BinaryNode* left = nullptr, BinaryNode* right = nullptr) {
    BinaryNode* node = new BinaryNode(value);
    node->left = left;
    node->right = right;
    return node;
}

int main() {
    // test: two null trees
    assert(flipEquivalentTrees(nullptr, nullptr) == true);

    // test: one null tree
    assert(flipEquivalentTrees(makeFETNode(1), nullptr) == false);
    assert(flipEquivalentTrees(nullptr, makeFETNode(1)) == false);

    // test: identical trees
    assert(flipEquivalentTrees(
        makeFETNode(1, makeFETNode(2), makeFETNode(3)),
        makeFETNode(1, makeFETNode(2), makeFETNode(3))) == true);

    // test: flipped at root
    assert(flipEquivalentTrees(
        makeFETNode(1, makeFETNode(2), makeFETNode(3)),
        makeFETNode(1, makeFETNode(3), makeFETNode(2))) == true);

    // test: different root values
    assert(flipEquivalentTrees(makeFETNode(1), makeFETNode(2)) == false);

    // test: different leaf values
    assert(flipEquivalentTrees(
        makeFETNode(1, makeFETNode(2), makeFETNode(3)),
        makeFETNode(1, makeFETNode(9), makeFETNode(3))) == false);

    std::cout << "All tests passed!" << std::endl;
    return 0;
}
